File: frontend/src/api/cash.ts
import apiClient from './client'

export type CashMovementType = 'in' | 'out'
export type CashEntrySource = 'manual' | 'deposit' | 'payment' | 'system_opening'

export interface LinkedAccountingEntry {
  id: number
  account_number: string
  label: string
  debit: string
  credit: string
}

export interface CashEntryConnections {
  can_delete: boolean
  blocking_reason: string | null
  accounting_entries: LinkedAccountingEntry[]
}

export interface CashEntry {
  id: number
  date: string
  amount: string
  type: CashMovementType
  contact_id: number | null
  payment_id: number | null
  reference: string | null
  description: string
  source: CashEntrySource
  balance_after: string
  is_system_opening: boolean
}

export interface CashEntryCreate {
  date: string
  amount: string
  type: CashMovementType
  contact_id?: number | null
  payment_id?: number | null
  reference?: string | null
  description?: string
}

export interface CashEntryUpdate {
  date?: string
  amount?: string
  type?: CashMovementType
  contact_id?: number | null
  payment_id?: number | null
  reference?: string | null
  description?: string | null
}

export interface CashCount {
  id: number
  date: string
  count_100: number
  count_50: number
  count_20: number
  count_10: number
  count_5: number
  count_2: number
  count_1: number
  count_cents_50: number
  count_cents_20: number
  count_cents_10: number
  count_cents_5: number
  count_cents_2: number
  count_cents_1: number
  total_counted: string
  balance_expected: string
  difference: string
  notes: string | null
}

export interface CashCountCreate {
  date: string
  count_100?: number
  count_50?: number
  count_20?: number
  count_10?: number
  count_5?: number
  count_2?: number
  count_1?: number
  count_cents_50?: number
  count_cents_20?: number
  count_cents_10?: number
  count_cents_5?: number
  count_cents_2?: number
  count_cents_1?: number
  notes?: string | null
}

export interface FundsChartRow {
  month: string
  balance: number
}

export async function getCashBalance(): Promise<{ balance: string }> {
  const response = await apiClient.get<{ balance: string }>('/api/cash/balance')
  return response.data
}

export async function getCashFundsChart(months = 6): Promise<FundsChartRow[]> {
  const response = await apiClient.get<FundsChartRow[]>('/api/cash/chart/funds', {
    params: { months },
  })
  return response.data
}

export async function listCashEntries(params?: {
  from_date?: string
  to_date?: string
  skip?: number
  limit?: number
}): Promise<CashEntry[]> {
  const response = await apiClient.get<CashEntry[]>('/api/cash/entries', { params })
  return response.data
}

export async function addCashEntry(payload: CashEntryCreate): Promise<CashEntry> {
  const response = await apiClient.post<CashEntry>('/api/cash/entries', payload)
  return response.data
}

export async function getCashEntry(id: number): Promise<CashEntry> {
  const response = await apiClient.get<CashEntry>(`/api/cash/entries/${id}`)
  return response.data
}

export async function updateCashEntry(id: number, payload: CashEntryUpdate): Promise<CashEntry> {
  const response = await apiClient.put<CashEntry>(`/api/cash/entries/${id}`, payload)
  return response.data
}

export async function deleteCashEntry(id: number): Promise<void> {
  await apiClient.delete(`/api/cash/entries/${id}`)
}

export async function getCashEntryConnections(id: number): Promise<CashEntryConnections> {
  const response = await apiClient.get<CashEntryConnections>(`/api/cash/entries/${id}/connections`)
  return response.data
}

export async function listCashCounts(params?: {
  from_date?: string
  to_date?: string
  skip?: number
  limit?: number
}): Promise<CashCount[]> {
  const response = await apiClient.get<CashCount[]>('/api/cash/counts', { params })
  return response.data
}

export async function addCashCount(payload: CashCountCreate): Promise<CashCount> {
  const response = await apiClient.post<CashCount>('/api/cash/counts', payload)
  return response.data
}

File: backend/schemas/cash.py
"""Pydantic schemas for cash register entries and cash counts."""

from __future__ import annotations

from datetime import date as _Date
from decimal import Decimal as _Decimal

from pydantic import BaseModel, field_validator

from backend.models.cash import CashEntrySource, CashMovementType


class CashEntryCreate(BaseModel):
    date: _Date
    amount: _Decimal
    type: CashMovementType
    contact_id: int | None = None
    payment_id: int | None = None
    reference: str | None = None
    description: str = ""

    @field_validator("amount")
    @classmethod
    def amount_positive(cls, v: _Decimal) -> _Decimal:
        if v <= 0:
            raise ValueError("amount must be positive")
        return v


class CashEntryUpdate(BaseModel):
    date: _Date | None = None
    amount: _Decimal | None = None
    type: CashMovementType | None = None
    contact_id: int | None = None
    payment_id: int | None = None
    reference: str | None = None
    description: str | None = None

    @field_validator("amount")
    @classmethod
    def amount_positive(cls, v: _Decimal | None) -> _Decimal | None:
        if v is not None and v <= 0:
            raise ValueError("amount must be positive")
        return v


class CashEntryRead(BaseModel):
    id: int
    date: _Date
    amount: _Decimal
    type: CashMovementType
    contact_id: int | None
    payment_id: int | None
    reference: str | None
    description: str
    source: CashEntrySource
    balance_after: _Decimal
    is_system_opening: bool

    model_config = {"from_attributes": True}


class LinkedAccountingEntry(BaseModel):
    """Minimal accounting entry info returned in cash connection checks."""

    id: int
    account_number: str
    label: str
    debit: _Decimal
    credit: _Decimal

    model_config = {"from_attributes": True}


class CashEntryConnectionsRead(BaseModel):
    """Result of checking what is connected to a cash entry before deletion."""

    can_delete: bool
    blocking_reason: str | None = None
    accounting_entries: list[LinkedAccountingEntry] = []


class CashCountCreate(BaseModel):
    date: _Date
    count_100: int = 0
    count_50: int = 0
    count_20: int = 0
    count_10: int = 0
    count_5: int = 0
    count_2: int = 0
    count_1: int = 0
    count_cents_50: int = 0
    count_cents_20: int = 0
    count_cents_10: int = 0
    count_cents_5: int = 0
    count_cents_2: int = 0
    count_cents_1: int = 0
    notes: str | None = None

    @field_validator(
        "count_100",
        "count_50",
        "count_20",
        "count_10",
        "count_5",
        "count_2",
        "count_1",
        "count_cents_50",
        "count_cents_20",
        "count_cents_10",
        "count_cents_5",
        "count_cents_2",
        "count_cents_1",
        mode="before",
    )
    @classmethod
    def counts_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("count must be non-negative")
        return v


class CashCountRead(BaseModel):
    id: int
    date: _Date
    count_100: int
    count_50: int
    count_20: int
    count_10: int
    count_5: int
    count_2: int
    count_1: int
    count_cents_50: int
    count_cents_20: int
    count_cents_10: int
    count_cents_5: int
    count_cents_2: int
    count_cents_1: int
    total_counted: _Decimal
    balance_expected: _Decimal
    difference: _Decimal
    notes: str | None

    model_config = {"from_attributes": True}


class CashBalanceRead(BaseModel):
    balance: _Decimal
